pkg/oci/oci.go

```diff
@@ -342,28 +350,48 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
 	var cfg *configfile.ConfigFile
 	var err error
 
-	authFileReader, err := os.Open(authOptions.AuthFile)
-	if err != nil {
-		// If the AuthFile was not set explicitly, we allow to fall back to the docker auth,
-		// otherwise we fail to avoid masking an error from the user
-		if !errors.Is(err, os.ErrNotExist) || authOptions.AuthFile != DefaultAuthFile {
-			return nil, fmt.Errorf("opening auth file %q: %w", authOptions.AuthFile, err)
+	if authOptions.PullSecret {
+		k8sClient, err := k8sutil.NewClientset("")
+		if err != nil {
+			return nil, fmt.Errorf("creating new k8s clientset: %w", err)
+		}
+		gps, err := k8sClient.CoreV1().Secrets("gadget").Get(context.TODO(), gadgetPullSecret, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("getting secret %q: %w", gadgetPullSecret, err)
+		}
+		if gps.Type != corev1.SecretTypeDockerConfigJson {
+			return nil, fmt.Errorf("secret %q is not of type %q", gadgetPullSecret, corev1.SecretTypeDockerConfigJson)
 		}
 
-		log.Debugf("Couldn't find default auth file %q...", authOptions.AuthFile)
-		log.Debugf("Using default docker auth file instead")
-		log.Debugf("$HOME: %q", os.Getenv("HOME"))
-
-		cfg, err = config.Load("")
+		cfg, err = config.LoadFromReader(bytes.NewReader(gps.Data[corev1.DockerConfigJsonKey]))
 		if err != nil {
 			return nil, fmt.Errorf("loading auth config: %w", err)
 		}
 	} else {
-		defer authFileReader.Close()
-		cfg, err = config.LoadFromReader(authFileReader)
+		authFileReader, err := os.Open(authOptions.AuthFile)
 		if err != nil {
-			return nil, fmt.Errorf("loading auth config: %w", err)
+			// If the AuthFile was not set explicitly, we allow to fall back to the docker auth,
+			// otherwise we fail to avoid masking an error from the user
+			if !errors.Is(err, os.ErrNotExist) || authOptions.AuthFile != DefaultAuthFile {
+				return nil, fmt.Errorf("opening auth file %q: %w", authOptions.AuthFile, err)
+			}
+
+			log.Debugf("Couldn't find default auth file %q...", authOptions.AuthFile)
+			log.Debugf("Using default docker auth file instead")
+			log.Debugf("$HOME: %q", os.Getenv("HOME"))
+
+			cfg, err = config.Load("")
+			if err != nil {
+				return nil, fmt.Errorf("loading auth config: %w", err)
+			}
+		} else {
+			defer authFileReader.Close()
+			cfg, err = config.LoadFromReader(authFileReader)
+			if err != nil {
+				return nil, fmt.Errorf("loading auth config: %w", err)
+			}
 		}
```

Comment:
I agree we could make it more readable but I was avoiding changing the way we check for falling back to docker auth like `en(authOptions.AuthFile) > 0` is always be true because of default path. How about following?

```
if authOptions.PullSecret {
	k8sClient, err := k8sutil.NewClientset("")
	if err != nil {
		return nil, fmt.Errorf("creating new k8s clientset: %w", err)
	}
	gps, err := k8sClient.CoreV1().Secrets("gadget").Get(context.TODO(), gadgetPullSecret, metav1.GetOptions{})
	if err != nil {
		return nil, fmt.Errorf("getting secret %q: %w", gadgetPullSecret, err)
	}
	if gps.Type != corev1.SecretTypeDockerConfigJson {
		return nil, fmt.Errorf("secret %q is not of type %q", gadgetPullSecret, corev1.SecretTypeDockerConfigJson)
	}

	cfg, err = config.LoadFromReader(bytes.NewReader(gps.Data[corev1.DockerConfigJsonKey]))
	if err != nil {
		return nil, fmt.Errorf("loading auth config: %w", err)
	}
} else if authFileReader, err := os.Open(authOptions.AuthFile); err != nil {
	// If the AuthFile was not set explicitly, we allow to fall back to the docker auth,
	// otherwise we fail to avoid masking an error from the user
	if !errors.Is(err, os.ErrNotExist) || authOptions.AuthFile != DefaultAuthFile {
		return nil, fmt.Errorf("opening auth file %q: %w", authOptions.AuthFile, err)
	}

	log.Debugf("Couldn't find default auth file %q...", authOptions.AuthFile)
	log.Debugf("Using default docker auth file instead")
	log.Debugf("$HOME: %q", os.Getenv("HOME"))

	cfg, err = config.Load("")
	if err != nil {
		return nil, fmt.Errorf("loading auth config: %w", err)
	}

} else {
	defer authFileReader.Close()
	cfg, err = config.LoadFromReader(authFileReader)
	if err != nil {
		return nil, fmt.Errorf("loading auth config: %w", err)
	}
}
```